File: lab1/lexer.c
/**********************************************************************/
/*Per Emilsson och Kenny Pettersson                                   */
/**********************************************************************/

/**********************************************************************/
/* lab 1 DVG C01 - Lexer OBJECT                                       */
/**********************************************************************/

/**********************************************************************/
/* Include files                                                      */
/**********************************************************************/
#include <stdio.h>
#include <stdbool.h>
#include <ctype.h>
#include <string.h>

/**********************************************************************/
/* Other OBJECT's METHODS (IMPORTED)                                  */
/**********************************************************************/
#include "keytoktab.h"

/**********************************************************************/
/* OBJECT ATTRIBUTES FOR THIS OBJECT (C MODULE)                       */
/**********************************************************************/
#define BUFSIZE 1024
#define LEXSIZE   30
static char buffer[BUFSIZE];
static char lexbuf[LEXSIZE];
static void pbuffer();
static int  pbuf  = 0;               /* current index program buffer  */
static int  plex  = 0;               /* current index lexeme  buffer  */
bool cutoff = false;
bool firstrun = true;


/**********************************************************************/
/*  PRIVATE METHODS for this OBJECT  (using "static" in C)            */
/**********************************************************************/
/**********************************************************************/
/* buffer functions                                                   */
/**********************************************************************/
/**********************************************************************/
/* Read the input file into the buffer                                */
/**********************************************************************/

static void get_prog()
{
	char temp;
	firstrun = false;
	//read stream from stdin, in our case > testfilexxx.pas

	//works
    while((temp = fgetc(stdin)) != EOF) {
		if(pbuf < BUFSIZE - 1)
			buffer[pbuf++] = temp;
		else
			break;

	}
	buffer[pbuf] = '$';
	pbuf = 0;
	if(pbuf == 9999)
		pbuffer();
}


/**********************************************************************/
/* Display the buffer                                                 */
/**********************************************************************/

static void pbuffer()
{
    for(int i = pbuf; i < sizeof(buffer);i++){
		printf("%c" ,buffer[i]);
	}
}

/**********************************************************************/
/* Copy a character from the program buffer to the lexeme buffer      */
/**********************************************************************/

static void get_char()
{
	//clear any whitespaces
	while(isspace(buffer[pbuf])){
		pbuf++;
	}
	lexbuf[plex++] = buffer[pbuf++];
}

/**********************************************************************/
/* End of buffer handling functions                                   */
/**********************************************************************/

/**********************************************************************/
/*  PUBLIC METHODS for this OBJECT  (EXPORTED)                        */
/**********************************************************************/
/**********************************************************************/
/* Return a token                                                     */
/**********************************************************************/
int get_token()
{
	if(firstrun) {
		get_prog();
	}
	plex = 0;							//reset plex 
	memset(lexbuf,0,sizeof(lexbuf));    //reset the lexbuffer at the start of every call to get_token
	get_char(); 						//get first char

    
	while (isalnum(buffer[pbuf])) {
		//handle signs 
        if(!isalnum(buffer[pbuf - 1]) && !isspace(buffer[pbuf - 1])){
			break;
		}
		//handle variables starting with number
		if(isdigit(buffer[pbuf-1]) && (isalpha(buffer[pbuf]))){
			break;
		}
		get_char();
	}
	//handle assign
	if(buffer[pbuf -1] == ':' && buffer[pbuf] == '='){
			get_char();
	}
	return lex2tok(lexbuf);
}



/**********************************************************************/
/* Return a lexeme                                                    */
/**********************************************************************/
char * get_lexeme()
{
	return lexbuf;
}

/**********************************************************************/
/* End of code                                                        */
/**********************************************************************/

File: lab4/lexer.py
# **********************************************************************
# *Per Emilsson och Kenny Pettersson                                   *
# **********************************************************************

import sys

from keytoktab import *

BUFSIZE = 1024

class Lexer:
    def __init__(self, bufsize: int = BUFSIZE):
        self.bufsize = bufsize
        self.source: str = ""
        self.pbuf: int = 0
        self.last_lexeme: str = ""
    
    def get_prog(self):
        data = sys.stdin.read()
        self.source = data[:self.bufsize-1] + '$' #slice of everything beyond the bufsize and add an $ to the end
        self.pbuf = 0
        self.pbuffer()

    #Display the buffer
    def pbuffer(self):
        # print everything except the final '$'
        sys.stdout.write(self.source[:-1])
        sys.stdout.flush()

    def get_char(self) -> str | None:
        # skip whitespace
        while self.pbuf < len(self.source) and self.source[self.pbuf].isspace():
            self.pbuf += 1
        if self.pbuf >= len(self.source):
            return None
        ch = self.source[self.pbuf]
        self.pbuf += 1
        return ch

    def get_token(self) -> int | str:
        if not self.source:
            self.get_prog()

        # build up the lexeme
        lexeme_chars = []
        ch = self.get_char()
        if ch is None:
            return error
        lexeme_chars.append(ch)

        # consume alnum sequence
        while self.pbuf < len(self.source) and self.source[self.pbuf].isalnum():
            prev, curr = self.source[self.pbuf-1], self.source[self.pbuf]
            if not (prev.isalnum() or prev.isspace()):
                break
            if prev.isdigit() and curr.isalpha():
                break
            lexeme_chars.append(curr)
            self.pbuf += 1

        # handle ':='
        if self.pbuf < len(self.source)-1 and lexeme_chars and lexeme_chars[-1] == ':' and self.source[self.pbuf] == '=':
            lexeme_chars.append('=')
            self.pbuf += 1

        # finalize the lexeme
        self.last_lexeme = ''.join(lexeme_chars)
        return lex2tok(self.last_lexeme)

    def get_lexeme(self) -> str:
        return self.last_lexeme